algorithms/pager.go

```diff
@@ -0,0 +1,52 @@
+package algorithms
+
+type FetchPageFunc[T any] func(page *string, size int64) ([]T, *PageInfo, error)
+
+type PageInfo struct {
+	HasNext  bool
+	After    *string
+	PageSize int64
+}
+
+type Pager[T any] struct {
+	pageInfo *PageInfo
+	err      error
+
+	fetchPage FetchPageFunc[T]
+}
+
+type PageIter[T any] interface {
+	NextPage() ([]T, error)
+}
+
+func NewPager[T any](pageSize int64, fetchPage FetchPageFunc[T]) *Pager[T] {
+	p := &Pager[T]{
+		fetchPage: fetchPage,
+	}
+	p.pageInfo = &PageInfo{
+		HasNext:  true,
+		After:    nil,
+		PageSize: pageSize,
+	}
+	return p
+}
+
+func (p *Pager[T]) NextPage() ([]T, error) {
+	var list []T
+
+	if p.err != nil {
+		return nil, p.err
+	}
+
+	if !p.pageInfo.HasNext {
+		return nil, nil
+	}
+
+	list, p.pageInfo, p.err = p.fetchPage(p.pageInfo.After, p.pageInfo.PageSize)
```

Comment:
Actually shouldn't we just return the error here directly as well instead of waiting until the next loop?

Reply:
yes, definitely